src/main/scala/firrtl/passes/InferResets.scala

```diff
@@ -0,0 +1,200 @@
+  private def analyze(c: Circuit): Map[ReferenceTarget, List[ResetDriver]] = {
+    type DriverMap = mutable.HashMap[ReferenceTarget, List[ResetDriver]]
+    def onMod(mod: DefModule): DriverMap = {
+      val instMap = mutable.Map[String, String]()
+      // We need to convert submodule port targets into targets on the Module port itself
+      def makeTarget(expr: Expression): ReferenceTarget = {
+        val target = toTarget(c.main, mod.name)(expr)
+        Utils.kind(expr) match {
+          case InstanceKind =>
+            val mod = instMap(target.ref)
+            val port = target.component.head match {
+              case TargetToken.Field(name) => name
+            }
+            target.copy(module = mod, ref = port, component = target.component.tail)
+          case _ => target
+        }
+      }
+      def onStmt(map: DriverMap)(stmt: Statement): Unit = {
+        stmt match {
+          case Connect(_, lhs, rhs) if lhs.tpe == ResetType =>
+            val target = makeTarget(lhs)
+            val driver = rhs.tpe match {
+              case ResetType => TargetDriver(makeTarget(rhs))
+              case tpe       => TypeDriver(tpe, () => makeTarget(rhs))
+            }
+            map(target) = driver :: Nil
+          case WDefInstance(_, inst, module, _) =>
+            instMap += (inst -> module)
+          case Conditionally(_, _, con, alt) =>
+            val conMap = new DriverMap
+            val altMap = new DriverMap
+            onStmt(conMap)(con)
+            onStmt(altMap)(alt)
+            // Default to outerscope if not found in alt
+            val altLookup = altMap.orElse(map).lift
+            for (key <- conMap.keys ++ altMap.keys) {
+              map(key) = (conMap.get(key).toList ++ altLookup(key).toList).flatten
+            }
+          case other => other.foreach(onStmt(map))
+        }
+      }
+      val types = new DriverMap
+      mod.foreach(onStmt(types))
+      types
+    }
+    c.modules.foldLeft(Map[ReferenceTarget, List[ResetDriver]]()) {
+      case (map, mod) => map ++ onMod(mod)
+    }
+  }
+
+  // Determine the type driving a given ResetType
+  private def resolve(map: Map[ReferenceTarget, List[ResetDriver]]): Try[Map[ReferenceTarget, Type]] = {
+    val res = mutable.Map[ReferenceTarget, Type]()
+    val errors = new Errors
+    def rec(target: ReferenceTarget): Type = {
+      val drivers = map(target)
+      res.getOrElseUpdate(target, {
+        val tpes = drivers.map {
+          case TargetDriver(t) => TypeDriver(rec(t), () => t)
+          case td: TypeDriver => td
+        }.groupBy(_.tpe)
+        if (tpes.keys.size != 1) {
+          // Multiple types of driver!
+          errors.append(DifferingDriverTypesException(target, tpes.toSeq))
+        }
+        tpes.keys.head
+      })
+    }
+    for ((target, _) <- map) {
+      rec(target)
+    }
+    Try {
+      errors.trigger()
+      res.toMap
+    }
+  }
+
+  private sealed trait TypeTree
```

Comment:
Add comment describing the high level purpose of these.

Reply:
done